### PR TITLE
Error on unused awaitable expressions

### PR DESCRIPTION
**Pull request by jhance**
Generates an error when an expression has a type which has a
defined __await__ in its MRO but is not used. This includes all
builtin awaitable objects (e.g. futures, coroutines, and tasks)
but also anything a user might define which is also awaitable.

A hint is attached to suggest awaiting.

This can be extended in the future to other types of values that
may lead to a resource leak if needed, or even exposed as a plugin.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
This one is a false positive: https://github.com/aiortc/aiortc/blob/c258ef4167eb04914646db03c6044ca05428893a/src/aiortc/rtcsctptransport.py#L1393. There's no need to await `asyncio.ensure_future`. And in any case, you can't add an await in a non-async function.

**Comment by jhance** [comment]
Failure to use a future likely indicates a resource leak bug because you won't cancel in the event of failure, so I wouldn't classify it as a false positive.

**Comment by jhance** [comment]
Also, docs explicitly say not to throw away the result of `ensure_future` or `create_task`: https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task

**Comment by JukkaL** [comment]
At least homeassistant seems to have some false positives. It has a `async_create_task` method that returns a `Task`, but it also records the task object so that actually won't be leaked, so the return value can be ignored. 

I can suggest a few things we could do about these:
1. Only complain about unused `Task` (or generally non-coroutine) values from certain known asyncio methods whose return values should not be ignored.
2. Use different error codes for ignored coroutines and other awaitables. Only complain about coroutines by default, and the general awaitable errors would have to be explicitly enabled, since they can cause false positives.

In any case, it would be good to use a dedicated error code for these errors.

There's also a self-check error:
```
mypy/checker.py:3442: error: Never apply isinstance() to unexpanded types; use
mypy.types.get_proper_type() first  [misc]
            if isinstance(typ, Instance):
               ^
```

**Comment by jhance** [comment]
Feature is split into two error codes so that the more controversial/harder one (unused task/future) is not on by default but unused coroutine, which everyone should want, is on by default.

File: mypy/errorcodes.py [hunk 1]
```diff
@@ -97,6 +97,9 @@ def __str__(self) -> str:
 LITERAL_REQ: Final = ErrorCode(
     "literal-required", "Check that value is a literal", 'General'
 )
+UNUSED_COROUTINE: Final = ErrorCode(
+    "unused-coroutine", "Ensure that all coroutines are used", "Genera"
```
Comment (JukkaL) [inline mypy/errorcodes.py:101]:
```suggestion
    "unused-coroutine", "Ensure that all coroutines are used", "General"
```

File: mypy/errorcodes.py [hunk 2]
```diff
@@ -147,6 +150,12 @@ def __str__(self) -> str:
     "General",
     default_enabled=False,
 )
+UNUSED_AWAITABLE: Final = ErrorCode(
+    "unused-awaitable",
+    "Ensure that all coroutines are used",
```
Comment (JukkaL) [inline mypy/errorcodes.py:155]: Should the description say something like "all awaitable values"?

**Review by JukkaL — APPROVED** [review]
Looks good now! Left two small comments/suggestions. Feel free to merge after you've addressed them.

**Pull request merged**